packages/coding-standard/src/Rules/NoScalarAndArrayConstructorParameterRule.php

```diff
@@ -23,99 +20,70 @@ final class NoScalarAndArrayConstructorParameterRule extends AbstractManyNodeTyp
+    private const VALUE_OBJECT_REGEX = '#\bValueObject\b#';
+
+    /**
+     * @var VariableAsParamAnalyser
+     */
+    private $variableAsParamTypeAnalyser;
+
+    /**
+     * @var ScalarTypeAnalyser
+     */
+    private $scalarTypeAnalyser;
+
+    public function __construct(
+        VariableAsParamAnalyser $variableAsParamTypeAnalyser,
+        ScalarTypeAnalyser $scalarTypeAnalyser
+    ) {
+        $this->variableAsParamTypeAnalyser = $variableAsParamTypeAnalyser;
+        $this->scalarTypeAnalyser = $scalarTypeAnalyser;
+    }
 
     /**
      * @return string[]
      */
     public function getNodeTypes(): array
     {
-        return [ClassMethod::class];
+        return [Variable::class];
     }
 
     /**
-     * @param ClassMethod $node
+     * @param Variable $node
      * @return string[]
      */
     public function process(Node $node, Scope $scope): array
     {
-        $parent = $node->getAttribute('parent');
-        if (! $parent instanceof Class_) {
+        if ($this->isValueObjectNamespace($scope)) {
             return [];
         }
 
-        if ($this->isValueObjectNamespace($parent)) {
+        $functionReflection = $scope->getFunction();
+        if (! $functionReflection instanceof MethodReflection) {
             return [];
         }
 
-        $methodName = (string) $node->name;
-        if (! $node->isMagic() || $methodName !== '__construct') {
+        if (! $this->variableAsParamTypeAnalyser->isVariableFromConstructorParam($functionReflection, $node)) {
             return [];
         }
 
-        $parameters = $node->getParams();
-        foreach ($parameters as $parameter) {
-            /** @var Identifier|Name|UnionType|NullableType|null $type */
-            $type = $parameter->type;
-
-            if ($type === null) {
-                continue;
-            }
-
-            $possibleTypes = $this->getPossibleTypes($type);
-            if ($this->isScalarOrArray($possibleTypes)) {
-                return [self::ERROR_MESSAGE];
-            }
-        }
-
-        return [];
-    }
-
-    private function isValueObjectNamespace(Class_ $class): bool
-    {
-        $namespacedName = $class->namespacedName->toString();
-        /** @var Identifier $name */
-        $name = $class->name;
-        /** @var string $className */
-        $className = $name->toString();
-        $pattern = sprintf(self::REGEX_VALUE_OBJECT_NAMESPACE_PLACEHOLDER, $className);
-
-        return Strings::match($namespacedName, $pattern) !== null;
-    }
-
-    /**
-     * @return mixed[]
-     */
-    private function getPossibleTypes($type): array
-    {
-        if ($type instanceof NullableType) {
-            return [$type->type];
-        }
-
-        if ($type instanceof UnionType) {
-            return $type->types;
+        // is variable in parameter?
+        $variableType = $scope->getType($node);
+        if (! $this->scalarTypeAnalyser->isScalarOrArrayType($variableType)) {
+            return [];
         }
```

Comment:
@samsonasik In the end I realized that we need static types, so we can analyse also these:

```php
/**
 * @param string $some
 */
public function __construct($some)
{
}
```

Unfortunatelly, PHPStan does not support type for `Param`. Not sure why (well, they don't need it)... but the type is already known there.

It only supports it for the first `Variable` that uses it. So I had to do a bit refactoring. Now we use the variables, so it works well.